### ISSUE TITLE
[Feature]: separation of NWBFile formation and dataset configuration

### ISSUE THREAD

**Issue by bendichter** [catalystneuro/neuroconv#441]
### What would you like to see added to NeuroConv?

We are in the process of generalizing NeuroConv to support HDF5 and Zarr backends and found that there are many places in the code that use `H5DataIO`, which is causing problems for the Zarr backend. The proposed solution is to remove all of these `H5DataIO` calls and instead provide a separate interface to alter these settings later. This provides a separation between setting the values of the data and metadata and setting the storage options.

Here is a proposal based on our conversation and some follow-up conversations with the NWB team:

1. Remove all compression-specific args from the `run_conversion` method signatures and all `H5DataIO` and any other backend-specific code from the underlying methods. Within `run_conversion`, all other code in `run_conversion()` would go into a new method, named something like `assemble_nwb()`. For each large dataset (where an `H5DataIO` was previously used) we would instead add each container to a dict. Something like: data_interface.dataset_configs = {(container_id, field): dict(**compression_kwargs)}

2. The `BaseDataInterface` would have an additional optional `backend` arg. This can be called during init or can be set later with `data_interface.set_backend("hdf5|zarr")`, or can be inferred if an NWB file already exists at that location, or will be default "hdf5". `NWBConverter` would similarly have a backend arg that propagates to inner interfaces.

3. `BaseDataInterface.set_dataset_configs()` would take args of the same form  as `dataset_configs`: `**{(container_id, field): dict(**compression_kwargs)}`. Anything not provided would have default behavior for the chosen backend.

4. `BaseDataInterface.write()` would call the `NWB**IO.write()` command based on the backend chosen.

5. `BaseDataInterface.run_conversion()` would stick around for backwards compatibility, and would call `.assemble_nwb()` and then `.write()`, and would not allow for the user to specify storage options.

Thoughts on this approach?

### Do you have any interest in helping implement the feature?

No.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/catalystneuro/neuroconv/blob/main/.github/CODE_OF_CONDUCT.rst)
- [X] Have you ensured this bug was not already [reported](https://github.com/catalystneuro/neuroconv/issues)?

**Comment by bendichter** [comment]
This is closely related to https://github.com/catalystneuro/neuroconv/pull/398

and would depend on https://github.com/hdmf-dev/hdmf/issues/856

**Comment by bendichter** [comment]
@CodyCBakerPhD This is a good start. I see you are making the `prepare_for_write` call on the `write_nwbfile`, which already knows the backend, so you should be able to omit it within the context. I think we can combine `make_or_load_nwbfile` into `write_nwbfile` and combine `parse_for_write` with `write`. Something like:

```python
def run_conversion(...):
  with NeuroConvIO(..., backend="zarr") as neuroconv_io:
       nwbfile = converter.assemble_nwb(nwbfile)  # or anything else that 'assembles' the file structure
       neuroconv_io.write(nwbfile, filepath, dataset_configurations=...)


class NeuroConvIO():
    def write(nwbfile, filepath, dataset_configurations: Optional[dict] = None):
        for (container_id, field), dset_config in dataset_configurations.items():
            data = nwbfile.get_object(container_id).getattr(field)
            nwbfile.get_object(container_id).getattr(field) = set_dataio(data=data, **dset_config)
        
```

**Comment by h-mayorquin** [comment]
I like the idea of adding separate interface methods for both appending (`add_to_nwb`) and writing. In my view, having separate methods instead of a single `run_conversion` that writes or appends depending on the input is way cleaner and transparent. 

Looking at #455 it seems that this `run_conversion` will still work for  `NWBConveter` and the `ConverterPipe`. I think it would be a good idea to add a write method to those classes. I like the idea of having consistent methods between stand-alone interfaces and the converters such as in:

https://neuroconv.readthedocs.io/en/main/conversion_examples_gallery/combinations/spikeglx_and_phy.html

**Comment by bendichter** [comment]
@h-mayorquin yes, that's the way I am thinking about it. metadata is added via `add_to_nwbfile`, then `.write` handles data packing settings

**Comment by h-mayorquin** [comment]
All right, then it is just missing from the signature in the base class.

**Comment by bendichter** [comment]
Well actually that was done intentionally. Note that with the **kwargs, you don't really need to put metadata in that signature- it will work even if not explicitly defined in the abstract class. There are edge cases where interfaces actually don't need a `metadata` arg  at all, and if you put `metadata` in the abstract class then code inspectors complain that the downstream class methods do not match the abstract method signature.

**Comment by h-mayorquin** [comment]
I see.
I was mislead about the purpose because the kwargs here are named conversion_options. But this is probably temporary until we fully separate adding from writing.

**Comment by CodyCBakerPhD** [comment]
This is mostly done; only more thorough testing of Zarr output per interface remains

Good to close?
